Percolator/Percolator.tla

```diff
@@ -0,0 +1,416 @@
+    /\ key_lock' = [key_lock EXCEPT ![primary] = @ \ {[ts |-> start_ts, primary |-> primary]}]
+    /\ UNCHANGED <<key_data>>
+
+Start(c) ==
+  /\ client_state[c] = "init"
+  /\ next_ts' = next_ts + 1
+  /\ client_state' = [client_state EXCEPT ![c] = "working"]
+  /\ client_ts' = [client_ts EXCEPT ![c].start_ts = next_ts']
+  /\ UNCHANGED <<key_lock, key_data, key_write, client_key>>
+
+Get(c) ==
+  /\ client_state[c] = "working"
+  /\ IF canPrewrite(c)
+       THEN
+         /\ client_state' = [client_state EXCEPT ![c] = "prewriting"]
+         /\ UNCHANGED <<next_ts, key_lock, key_data, key_write, client_ts, client_key>>
+       ELSE
+         /\ cleanup(c)
+         /\ UNCHANGED <<next_ts, client_state, client_ts, client_key>>
+
+Prewrite(c) ==
+  /\ client_state[c] = "prewriting"
+  /\ IF canCommit(c)
+       THEN
+         /\ next_ts' = next_ts + 1
+         /\ client_state' = [client_state EXCEPT ![c] = "committing"]
+         /\ client_ts' = [client_ts EXCEPT ![c].commit_ts = next_ts']
+         /\ UNCHANGED <<key_lock, key_data, key_write, client_key>>
+       ELSE
+         /\ lock(c)
+         /\ UNCHANGED <<next_ts, client_state, client_ts, client_key>>
+
+Commit(c) ==
+  /\ client_state[c] = "committing"
+  /\ commitPrimary(c)
+  /\ client_state' = [client_state EXCEPT ![c] = "committed"]
+  /\ UNCHANGED <<next_ts, client_ts, client_key>>
+  \* If we commit the primary key successfully, we can think the transaction is
+  \* committed.
+
+ClientOp(c) ==
+  \/ Start(c)
+  \/ Get(c)
+  \/ Prewrite(c)
+  \/ Commit(c)
+
+Next == \E c \in CLIENT : ClientOp(c)
+
+Init ==
+  LET
+    \* Selects a primary key and use the rest for the secondary keys.
+    chooseKey(ks) ==
+    LET
+      primary == CHOOSE k \in ks : TRUE
+    IN
+      [primary |-> primary, secondary |-> ks \ {primary}]
+  IN
+    /\ next_ts = 0
+    /\ client_state = [c \in CLIENT |-> "init"]
+    /\ client_ts = [c \in CLIENT |-> [start_ts |-> 0, commit_ts |-> 0]]
+    /\ client_key = [c \in CLIENT |-> chooseKey(KEY)]
+    /\ key_lock = [k \in KEY |-> {}]
+    /\ key_write = [k \in KEY |-> <<>>]
+    /\ key_data = [k \in KEY |-> {}]
+
+PercolatorSpec == Init /\ [][Next]_vars
+
+NextTsTypeInv ==
+  next_ts \in Nat
+
+ClientStateTypeInv ==
+  client_state \in [CLIENT -> {"init", "working", "prewriting", "committing", "committed"}]
+
+ClientTsTypeInv ==
+  client_ts \in [CLIENT -> [start_ts : Nat, commit_ts : Nat]]
+
+ClientKeyTypeInv ==
+  client_key \in [CLIENT -> [primary : KEY, secondary: SUBSET KEY]]
+
+KeyDataTypeInv ==
+  key_data \in [KEY -> SUBSET Nat]
+
+KeyLockTypeInv ==
+  key_lock \in [KEY -> SUBSET [ts : Nat, primary : KEY]]
+
+KeyWriteTypeInv ==
+  key_write \in [KEY -> Seq([start_ts: Nat, commit_ts: Nat])]
+
+TypeInvariant ==
+  /\ NextTsTypeInv
+  /\ ClientStateTypeInv
+  /\ ClientTsTypeInv
+  /\ ClientKeyTypeInv
+  /\ KeyDataTypeInv
+  /\ KeyLockTypeInv
+  /\ KeyWriteTypeInv
+
+Invariant == TypeInvariant
+
+LEMMA InitStateSatisfiesInvariant ==
```

Comment:
can we move the proof to the end of the file?

Reply:
Done.